Lib/shutil.py

```diff
@@ -524,13 +524,15 @@ def _copytree(entries, src, dst, symlinks, ignore, copy_function,
                     else:
                         copy_function(srcobj, dstname)
             elif srcentry.is_dir():
+                # Will raise shutil.Error containing multiple exceptions gathered
+                # from recursion
                 copytree(srcobj, dstname, symlinks, ignore, copy_function,
                          ignore_dangling_symlinks, dirs_exist_ok)
             else:
-                # Will raise a SpecialFileError for unsupported file types
+                # Can raise SpecialFileError or SameFileError
                 copy_function(srcobj, dstname)
-        # catch the Error from the recursive copytree so that we can
-        # continue with other files
+        except SameFileError as err:
+            errors.append((srcname, dstname, str(err)))
         except Error as err:
             errors.extend(err.args[0])
```

Comment:
Would it instead be better to use:
```python
if type(e) is Error:
  errors.extend(err.args[0])
else:
  errors.append((srcname, dstname, str(err)))
```
which perhaps is more robust

Reply:
Yes, it is more robust if there are more bugs that we have not found. It is best if we could find them too and add a test case for them. 

If it is impossible to trigger other exceptions in the outermost recursion layer, I would prefer the code to be explicit and only catch what is attainable.

Reply:
> I would prefer the code to be explicit and only catch what is attainable.

Agreed; perhaps then we should add
```python
class ErrorGroup(Error): pass
```
and throw/catch this when grouping the exceptions. That way it is explicit that this merging behavior is actually going to do the right thinking, rather than hoping the exception is the one we think it is.

Reply:
Thanks, that looked good in the code! Fixed